File: Thyro-Care-main/src/types/knowledgeGovernance.ts
/**
 * Knowledge governance types aligned with backend enums/schemas (Phase 12).
 *
 * Clients never supply actor identity, statuses, or approval timestamps — those are
 * always derived server-side from the authenticated user and workflow transitions.
 */

/** `archived` and `in_review` are legacy statuses retained for backward compatibility. */
export type KnowledgeStatus =
  | "draft"
  | "in_review"
  | "pending_review"
  | "changes_requested"
  | "approved"
  | "rejected"
  | "archived"
  | "retired";

export type KnowledgeReviewDecision =
  | "approve"
  | "request_changes"
  | "reject"
  | "retire"
  | "restore";

export type KnowledgeLanguage = "english" | "sinhala" | "tamil";

export type KnowledgeTopic =
  | "thyroidectomy_recovery"
  | "thyroid_cancer_survivorship"
  | "medication_education"
  | "follow_up_education"
  | "symptom_awareness"
  | "nutrition_education"
  | "emergency_awareness"
  | "general_education"
  | "other";

export const KNOWLEDGE_STATUS_OPTIONS: { value: KnowledgeStatus; label: string }[] = [
  { value: "draft", label: "Draft" },
  { value: "pending_review", label: "Pending review" },
  { value: "changes_requested", label: "Changes requested" },
  { value: "approved", label: "Approved" },
  { value: "rejected", label: "Rejected" },
  { value: "retired", label: "Retired" },
];

export const KNOWLEDGE_TOPIC_OPTIONS: { value: KnowledgeTopic; label: string }[] = [
  { value: "thyroidectomy_recovery", label: "Thyroidectomy recovery" },
  { value: "thyroid_cancer_survivorship", label: "Thyroid cancer survivorship" },
  { value: "medication_education", label: "Medication education" },
  { value: "follow_up_education", label: "Follow-up education" },
  { value: "symptom_awareness", label: "Symptom awareness" },
  { value: "nutrition_education", label: "Nutrition education" },
  { value: "emergency_awareness", label: "Emergency awareness" },
  { value: "general_education", label: "General education" },
  { value: "other", label: "Other" },
];

export const KNOWLEDGE_LANGUAGE_OPTIONS: { value: KnowledgeLanguage; label: string }[] = [
  { value: "english", label: "English" },
  { value: "sinhala", label: "Sinhala" },
  { value: "tamil", label: "Tamil" },
];

export interface KnowledgeVersion {
  version_id: string;
  version_number: number;
  title: string;
  source_name: string;
  source_url: string | null;
  topic: string;
  language: string;
  body: string;
  medical_disclaimer: string;
  content_hash: string;
  review_status: KnowledgeStatus;
  created_by_user_id: string | null;
  created_at: string;
  submitted_for_review_at: string | null;
  submitted_by_user_id: string | null;
  approved_at: string | null;
  approved_by_user_id: string | null;
  rejected_at: string | null;
  rejected_by_user_id: string | null;
  retired_at: string | null;
  retired_by_user_id: string | null;
  review_summary: string | null;
  supersedes_version_id: string | null;
  version: number;
}

export interface KnowledgeDocument {
  document_id: string;
  slug: string;
  title: string;
  current_version_id: string | null;
  current_version_number?: number | null;
  current_status: KnowledgeStatus;
  topic: string;
  language: string;
  created_by_user_id: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeDocumentDetail {
  document: KnowledgeDocument;
  current_version: KnowledgeVersion | null;
  versions: KnowledgeVersion[];
}

export interface KnowledgeDocumentListResponse {
  items: KnowledgeDocument[];
  total: number;
  page: number;
  page_size: number;
}

export interface KnowledgeReviewQueueItem {
  document_id: string;
  version_id: string;
  version_number: number;
  title: string;
  topic: string;
  language: string;
  content_hash: string;
  submitted_for_review_at: string | null;
  submitted_by_user_id: string | null;
}

export interface KnowledgeReviewQueueResponse {
  items: KnowledgeReviewQueueItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface KnowledgeReviewRecord {
  id: string;
  document_id: string;
  version_id: string;
  reviewer_user_id: string;
  reviewer_role: string;
  decision: KnowledgeReviewDecision;
  reviewed_content_hash: string;
  comments: string | null;
  created_at: string;
}

export type IngestionStatus = "completed" | "failed";

export interface KnowledgeApprovalResult {
  document: KnowledgeDocument;
  version: KnowledgeVersion;
  ingestion_status: IngestionStatus | string;
  ingestion_message: string | null;
}

export interface KnowledgeDiffLine {
  op: string;
  from_line: string | null;
  to_line: string | null;
}

export interface KnowledgeCompareResponse {
  document_id: string;
  from_version_id: string | null;
  to_version_id: string;
  lines: KnowledgeDiffLine[];
  truncated: boolean;
}

// ---- Requests -----------------------------------------------------------------------

export interface KnowledgeDraftCreateRequest {
  slug: string;
  title: string;
  source_name: string;
  source_url?: string | null;
  topic: KnowledgeTopic;
  language?: KnowledgeLanguage;
  body: string;
  medical_disclaimer?: string;
}

export interface KnowledgeDraftUpdateRequest {
  title?: string;
  source_name?: string;
  source_url?: string | null;
  topic?: KnowledgeTopic;
  language?: KnowledgeLanguage;
  body?: string;
  medical_disclaimer?: string;
  expected_version: number;
}

export interface KnowledgeSubmitRequest {
  expected_version: number;
  submission_note?: string | null;
}

export interface KnowledgeReviewDecisionRequest {
  decision: KnowledgeReviewDecision;
  expected_version: number;
  expected_content_hash: string;
  comments?: string | null;
}

export interface KnowledgeNewVersionRequest {
  expected_version: number;
  title?: string;
  source_name?: string;
  source_url?: string | null;
  topic?: KnowledgeTopic;
  language?: KnowledgeLanguage;
  body?: string;
  medical_disclaimer?: string;
}

export interface KnowledgeRetireRequest {
  expected_version: number;
  reason: string;
}

export interface KnowledgeRestoreRequest {
  expected_version: number;
  expected_content_hash: string;
}

export interface KnowledgeListParams {
  status?: KnowledgeStatus;
  topic?: string;
  language?: string;
  page?: number;
  page_size?: number;
}

File: Thyro-Care-main/backend/app/schemas/knowledge_governance.py
"""Knowledge governance request/response schemas (Phase 12).

Clients never supply actor identity, statuses, or approval timestamps — those are
always derived server-side from the authenticated user and workflow transitions.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from urllib.parse import urlparse

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import (
    KnowledgeLanguage,
    KnowledgeReviewDecision,
    KnowledgeStatus,
    KnowledgeTopic,
)
from app.models.knowledge import KNOWLEDGE_BODY_MAX, REVIEW_COMMENTS_MAX

_TITLE_MAX = 300
_SOURCE_NAME_MAX = 200
_URL_MAX = 500
_DISCLAIMER_MAX = 1000
_SLUG_MAX = 200
_REASON_MAX = 1000


def _empty_to_none(value: Any) -> Any:
    if isinstance(value, str) and value.strip() == "":
        return None
    return value


def _validate_url(value: str | None) -> str | None:
    if value is None:
        return None
    cleaned = value.strip()
    if not cleaned:
        return None
    parsed = urlparse(cleaned)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise ValueError("source_url must be an http or https URL")
    return cleaned


def _strip_required(value: str) -> str:
    cleaned = value.strip()
    if not cleaned:
        raise ValueError("This field is required")
    return cleaned


class KnowledgeDraftCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    slug: str = Field(min_length=1, max_length=_SLUG_MAX)
    title: str = Field(min_length=1, max_length=_TITLE_MAX)
    source_name: str = Field(min_length=1, max_length=_SOURCE_NAME_MAX)
    source_url: str | None = Field(default=None, max_length=_URL_MAX)
    topic: KnowledgeTopic
    language: KnowledgeLanguage = KnowledgeLanguage.ENGLISH
    body: str = Field(min_length=1, max_length=KNOWLEDGE_BODY_MAX)
    medical_disclaimer: str = Field(default="", max_length=_DISCLAIMER_MAX)

    @field_validator("slug", "title", "source_name", "body")
    @classmethod
    def _strip(cls, value: str) -> str:
        return _strip_required(value)

    @field_validator("source_url", mode="before")
    @classmethod
    def _blank_url(cls, value: Any) -> Any:
        return _empty_to_none(value)

    @field_validator("source_url")
    @classmethod
    def _check_url(cls, value: str | None) -> str | None:
        return _validate_url(value)


class KnowledgeDraftUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=_TITLE_MAX)
    source_name: str | None = Field(default=None, min_length=1, max_length=_SOURCE_NAME_MAX)
    source_url: str | None = Field(default=None, max_length=_URL_MAX)
    topic: KnowledgeTopic | None = None
    language: KnowledgeLanguage | None = None
    body: str | None = Field(default=None, min_length=1, max_length=KNOWLEDGE_BODY_MAX)
    medical_disclaimer: str | None = Field(default=None, max_length=_DISCLAIMER_MAX)
    expected_version: int = Field(ge=1)

    @field_validator("title", "source_name", "body")
    @classmethod
    def _strip_optional(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return _strip_required(value)

    @field_validator("source_url", mode="before")
    @classmethod
    def _blank_url(cls, value: Any) -> Any:
        return _empty_to_none(value)

    @field_validator("source_url")
    @classmethod
    def _check_url(cls, value: str | None) -> str | None:
        return _validate_url(value)

    def editable_payload(self) -> dict[str, Any]:
        return self.model_dump(exclude_unset=True, exclude={"expected_version"})


class KnowledgeSubmitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    submission_note: str | None = Field(default=None, max_length=REVIEW_COMMENTS_MAX)


class KnowledgeReviewDecisionRequest(BaseModel):
    """Unified medical-expert decision payload (approve / request changes / reject)."""

    model_config = ConfigDict(extra="forbid")

    decision: KnowledgeReviewDecision
    expected_version: int = Field(ge=1)
    expected_content_hash: str = Field(min_length=1, max_length=128)
    comments: str | None = Field(default=None, max_length=REVIEW_COMMENTS_MAX)

    @field_validator("comments", mode="before")
    @classmethod
    def _blank_comments(cls, value: Any) -> Any:
        return _empty_to_none(value)


class KnowledgeApproveRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    expected_content_hash: str = Field(min_length=1, max_length=128)
    review_summary: str | None = Field(default=None, max_length=REVIEW_COMMENTS_MAX)


class KnowledgeRequestChangesRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    expected_content_hash: str = Field(min_length=1, max_length=128)
    comments: str = Field(min_length=1, max_length=REVIEW_COMMENTS_MAX)


class KnowledgeRejectRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    expected_content_hash: str = Field(min_length=1, max_length=128)
    comments: str = Field(min_length=1, max_length=REVIEW_COMMENTS_MAX)


class KnowledgeNewVersionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    title: str | None = Field(default=None, min_length=1, max_length=_TITLE_MAX)
    source_name: str | None = Field(default=None, min_length=1, max_length=_SOURCE_NAME_MAX)
    source_url: str | None = Field(default=None, max_length=_URL_MAX)
    topic: KnowledgeTopic | None = None
    language: KnowledgeLanguage | None = None
    body: str | None = Field(default=None, min_length=1, max_length=KNOWLEDGE_BODY_MAX)
    medical_disclaimer: str | None = Field(default=None, max_length=_DISCLAIMER_MAX)

    @field_validator("source_url", mode="before")
    @classmethod
    def _blank_url(cls, value: Any) -> Any:
        return _empty_to_none(value)

    @field_validator("source_url")
    @classmethod
    def _check_url(cls, value: str | None) -> str | None:
        return _validate_url(value)

    def overrides(self) -> dict[str, Any]:
        return self.model_dump(exclude_unset=True, exclude={"expected_version"})


class KnowledgeRetireRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=1, max_length=_REASON_MAX)


class KnowledgeRestoreRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    expected_content_hash: str = Field(min_length=1, max_length=128)


class KnowledgeRetryIngestRequest(BaseModel):
    """Retry publication for an already-approved version without re-approving."""

    model_config = ConfigDict(extra="forbid")

    expected_content_hash: str = Field(min_length=1, max_length=128)


# ---- Responses --------------------------------------------------------------------


class KnowledgeVersionPublic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version_id: str
    version_number: int
    title: str
    source_name: str
    source_url: str | None = None
    topic: str
    language: str
    body: str
    medical_disclaimer: str
    content_hash: str
    review_status: KnowledgeStatus
    created_by_user_id: str | None = None
    created_at: datetime
    submitted_for_review_at: datetime | None = None
    submitted_by_user_id: str | None = None
    approved_at: datetime | None = None
    approved_by_user_id: str | None = None
    rejected_at: datetime | None = None
    rejected_by_user_id: str | None = None
    retired_at: datetime | None = None
    retired_by_user_id: str | None = None
    review_summary: str | None = None
    supersedes_version_id: str | None = None
    version: int


class KnowledgeDocumentPublic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document_id: str
    slug: str
    title: str = ""
    current_version_id: str | None = None
    current_version_number: int | None = None
    current_status: KnowledgeStatus
    topic: str
    language: str
    created_by_user_id: str | None = None
    version: int
    created_at: datetime
    updated_at: datetime


class KnowledgeDocumentDetail(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document: KnowledgeDocumentPublic
    current_version: KnowledgeVersionPublic | None = None
    versions: list[KnowledgeVersionPublic] = Field(default_factory=list)


class KnowledgeDocumentListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[KnowledgeDocumentPublic]
    total: int
    page: int
    page_size: int


class KnowledgeReviewQueueItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document_id: str
    version_id: str
    version_number: int
    title: str
    topic: str
    language: str
    content_hash: str
    submitted_for_review_at: datetime | None = None
    submitted_by_user_id: str | None = None


class KnowledgeReviewQueueResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[KnowledgeReviewQueueItem]
    total: int
    page: int
    page_size: int


class KnowledgeReviewRecordPublic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    document_id: str
    version_id: str
    reviewer_user_id: str
    reviewer_role: str
    decision: KnowledgeReviewDecision
    reviewed_content_hash: str
    comments: str | None = None
    created_at: datetime


class KnowledgeApprovalResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document: KnowledgeDocumentPublic
    version: KnowledgeVersionPublic
    ingestion_status: str
    ingestion_message: str | None = None


class KnowledgeDiffLine(BaseModel):
    model_config = ConfigDict(extra="forbid")

    op: str
    from_line: str | None = None
    to_line: str | None = None


class KnowledgeCompareResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document_id: str
    from_version_id: str | None = None
    to_version_id: str
    lines: list[KnowledgeDiffLine] = Field(default_factory=list)
    truncated: bool = False


class MessageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    message: str
